lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1039,55 +1231,473 @@ def _on_logging_relation_departed(self, event):
         self.on.loki_push_api_endpoint_departed.emit()
 
-    def _check_alert_rules(self, alert_groups, invalid_files) -> str:
-        """Check alert rules.
+    @property
+    def loki_push_api(self) -> List[str]:
+        """Fetch Loki Push API endpoints sent from LokiPushApiProvider through relation data.
+
+        Returns:
+            A list with Loki Push API endpoints.
+        """
+        return [
+            _type_convert_stored(loki_endpoint)
+            for loki_endpoint in self._stored.loki_push_api.values()
+        ]
+
+
+class PromtailDigestError(Exception):
+    """Raised if there is an error with Promtail binary file."""
+
+
+class LogProxyConsumer(RelationManagerBase):
+    """LogProxyConsumer class."""
+
+    _stored = StoredState()
+
+    def __init__(
+        self,
+        charm,
+        log_files: list,
+        container_name: Optional[str],
+        relation_name: str = "log_proxy",
+        enable_syslog: bool = False,
+        syslog_port: int = 1514,
+    ):
+        super().__init__(charm, relation_name)
+        self._stored.set_default(grafana_agents="{}")
+        self._charm = charm
+        self._relation_name = relation_name
+        self._container_name = container_name
+        self._container = self._get_container(container_name)
+        self._log_files = log_files or []
+        self._syslog_port = syslog_port
+        self._is_syslog = enable_syslog
+        self.framework.observe(
+            self._charm.on.log_proxy_relation_created, self._on_log_proxy_relation_created
+        )
+        self.framework.observe(
+            self._charm.on.log_proxy_relation_changed, self._on_log_proxy_relation_changed
+        )
+        self.framework.observe(
+            self._charm.on.log_proxy_relation_departed, self._on_log_proxy_relation_departed
+        )
+
+    def _on_log_proxy_relation_created(self, event):
+        """Event handler for the `log_proxy_relation_created`."""
+        self._create_directories()
+        self._container.push(
+            WORKLOAD_CONFIG_PATH, yaml.safe_dump(self._initial_config), make_dirs=True
+        )
+
+    def _on_log_proxy_relation_changed(self, event):
+        """Event handler for the `log_proxy_relation_changed`.
 
         Args:
-            alert_groups: a list of prometheus alert rule groups.
-            invalid_files: a list of invalid rules files.
+            event: The event object `RelationChangedEvent`.
+        """
+        if event.relation.data[event.app].get("data", None):
+            try:
+                self._obtain_promtail(event)
+                self._update_config(event)
+                self._update_agents_list(event)
+                self._add_pebble_layer()
+                self._container.restart(WORKLOAD_SERVICE_NAME)
+            except HTTPError as e:
+                msg = "Promtail binary couldn't be download - {}".format(str(e))
+                logger.warning(msg)
+                raise PromtailDigestError(msg)
```

Comment:
iiuc we can narrow the span of the try:

```suggestion
            try:
                self._obtain_promtail(event)
            except HTTPError as e:
                msg = "Promtail binary couldn't be download - {}".format(str(e))
                logger.warning(msg)
                raise PromtailDigestError(msg)
            else:
                self._update_config(event)
                self._update_agents_list(event)
                self._add_pebble_layer()
                self._container.restart(WORKLOAD_SERVICE_NAME)
```

Reply:
absolutely! Good catch!